### PR TITLE
BUG/API: DTI/TDI/PI.insert cast to object on failure

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

ATM DTI/TDI/PI/insert are pretty idiosyncratic.  This changes them to match the base Index.insert by a) always trying _validate_fill_value (this changes the behavior for strings that can be parsed to datetime/timedelta/period) and b) always fall back to object in cases where this validation fails (this changes behavior for non-strings where the validation fails).

AFAICT the current behavior was implemented in #5819 and the only discussion I see about castable strings or non-castable non-strings is this [comment](https://github.com/pandas-dev/pandas/pull/5819#issuecomment-31454545), which I don't think is relevant any longer.

One corner case is mismatched tzs, which raise in master.  In the PR they cast to object.  They could also reasonably cast to the index's tz (xref #37605)

cc @rhshadrach IIRC you had a question about the insert tests in test_coercion.  I think this addresses what you were asking about.  Let me know if I missed something.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
xref #37774

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -589,7 +589,11 @@ def delete(self, loc):
 
     @doc(NDArrayBackedExtensionIndex.insert)
     def insert(self, loc: int, item):
-        result = super().insert(loc, item)
+        try:
+            result = super().insert(loc, item)
+        except (ValueError, TypeError):
+            # i.e. self._data._validate_scalar raised
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py:595]: really? it seems that the base can handle this no?

**Review by jreback — CHANGES_REQUESTED** [review]
a more general question here. why are we trying to accomodate this at all with a non-conforming value? shouldn't we just plain raise? and leave it to the caller to `astype(object)`?

of course this might be a *moar* breaking change that what is here.

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py:595]: in the non-raising case we still need to do the _get_insert_freq a few lines down

**Comment by jbrockmendel** [comment]
> a more general question here. why are we trying to accomodate this at all with a non-conforming value? shouldn't we just plain raise? and leave it to the caller to astype(object)?

if we were starting from scratch that'd be a reasonable option, but implementing that consistently would break a lot of existing code.  The best we can do is try to be internally consistent.

**Comment by jorisvandenbossche** [comment]
> a more general question here. why are we trying to accomodate this at all with a non-conforming value? shouldn't we just plain raise? and leave it to the caller to astype(object)?

Agreed with this, and that is also why I have been arguing for this behaviour for the new extension dtypes. 

I certainly follow the goal to be consistent internally, but *if* we prefer the strict behavior, and *if* we do that for the new nullable extension dtypes (both "if"s to discuss), we will have some inconsistency for a period of time anyway. In which case it might make sense to *not* fix the remaining current inconsistencies as done here.

**Comment by jbrockmendel** [comment]
> why are we trying to accomodate this at all with a non-conforming value? shouldn't we just plain raise? and leave it to the caller to astype(object)?

This comes up in at least two fairly-common contexts:

1) `df[new_col] = foo` when `new_col` doesn't match `df.columns.dtype`
2) `obj.reset_index()`

It seems awkward to ask users to do `df.columns = df.columns.astype(object)` before that kind of operation, especially since the reset_index can happen somewhat deep inside other functions.

> and if we do that for the new nullable extension dtypes (both "if"s to discuss), we will have some inconsistency for a period of time anyway

I'm not sure i follow.  This is about `Index.insert`, so nullable dtypes are still aspirational.

**Comment by jorisvandenbossche** [comment]
Ah yes, sorry, I had "arrays" in my head when writing that, and not "indexes" (we still need to have the discussion for arrays, will need to open a dedicated issue about that at some point). 

Fully agreed that those use cases you mention are important for Index objects where automatic upcasting is useful.

> AFAICT the current behavior was implemented in #5819 and the only discussion I see about castable strings or non-castable non-strings is this comment, which I don't think is relevant any longer.

No longer relevant in the sense that we no longer check if a string is datetime-like ? Or ..?

> One corner case is mismatched tzs, which raise in master. In the PR they cast to object. They could also reasonably cast to the index's tz (xref #37605)

Yes, and similarly as being discussed on that issue, my preference would be to not cast to object for this case, but preserve the original tz

**Comment by jbrockmendel** [comment]
i think the earlier non-consensus was a misunderstanding and outstanding issues have been resolved?

**Review by jreback — CHANGES_REQUESTED** [review]
can you construct a whatsnew note about this. *maybe* should be a sub-section, though ok i think with just a listing of what specific types this will now upcast to object rather than raise.

**Comment by jbrockmendel** [comment]
whatsnew added + green

File: doc/source/whatsnew/v1.3.0.rst [hunk 2]
```diff
@@ -283,6 +283,7 @@ Indexing
 - Bug in :meth:`DataFrame.loc` dropping levels of :class:`MultiIndex` when :class:`DataFrame` used as input has only one row (:issue:`10521`)
 - Bug in setting ``timedelta64`` values into numeric :class:`Series` failing to cast to object dtype (:issue:`39086`)
 - Bug in setting :class:`Interval` values into a :class:`Series` or :class:`DataFrame` with mismatched :class:`IntervalDtype` incorrectly casting the new values to the existing dtype (:issue:`39120`)
+- Bug in incorrectly raising when setting a new column that cannot be held in the existing ``frame.columns`` instead of casting to a compatible dtype (:issue:`39068`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:286]: hmm isn't this directly `Index.insert` ?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.3.0.rst (hunk 2 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v1.3.0.rst:286]: Yes, could flesh this out (.reset_index is also affected)

Comment (jreback) [reply doc/source/whatsnew/v1.3.0.rst:286]: yeah i think its worth being more explict here as this is a non-trivial change

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**